packages/react-data-grid/src/utils/columnUtils.ts

```diff
@@ -8,21 +8,41 @@ type Metrics<R> = Pick<ColumnMetrics<R>, 'viewportWidth' | 'minColumnWidth' | 'c
 export function getColumnMetrics<R>(metrics: Metrics<R>): ColumnMetrics<R> {
   let left = 0;
   let totalWidth = 0;
+  let allocatedWidths = 0;
+  let unassignedColumnsCount = 0;
+  let lastFrozenColumnIndex = -1;
+  const columns: Column<R>[] = [];
+
+  for (const metricsColumn of metrics.columns) {
+    const column = { ...metricsColumn };
+    setSpecifiedWidth(column, metrics.columnWidths, metrics.viewportWidth, metrics.minColumnWidth);
+
+    if (column.width === undefined) {
+      unassignedColumnsCount++;
+    } else {
+      allocatedWidths += column.width;
+    }
 
-  const columns = metrics.columns.map(c => ({ ...c }));
-  setSpecifiedWidths(columns, metrics.columnWidths, metrics.viewportWidth, metrics.minColumnWidth);
+    if (isFrozen(column)) {
+      lastFrozenColumnIndex++;
+      columns.splice(lastFrozenColumnIndex, 0, column);
+    } else {
+      columns.push(column);
+    }
+  }
 
-  const allocatedWidths = columns.map(c => c.width || 0).reduce((acc, w) => acc + w, 0);
   const unallocatedWidth = metrics.viewportWidth - allocatedWidths - getScrollbarSize();
+  const unallocatedColumnWidth = Math.max(
+    Math.floor(unallocatedWidth / unassignedColumnsCount),
+    metrics.minColumnWidth
+  );
 
-  setRemainingWidths(columns, unallocatedWidth, metrics.minColumnWidth);
-
-  let calculatedColumns: CalculatedColumn<R>[] = columns.map(column => {
+  const calculatedColumns: CalculatedColumn<R>[] = columns.map((column, idx) => {
     // Every column should have a valid width as this stage
-    const width = column.width as number;
+    const width = column.width || unallocatedColumnWidth;
```

Comment:
This would be crazy that maybe no one would intentionally do that, but what if the user set `column.width` to `0`? Should we check based on `undefined` like above rather than `undefined null and 0`?